File: lesson-2_recursion-and-recurrence/the-longest-path/memoization/main.kt
import kotlin.math.max

fun main() {
    val n = readLine()!!.toInt()
    val t = Array(n) { emptyArray<Int>() }
    val rec = Array(n) { emptyArray<Int>() }
    for (i in 0 until n) {
        t[i] = Array(i + 1) { 0 }
        rec[i] = Array(i + 1) { -1 }

        val tmp = readLine()!!.split(' ')
        for ((j, v) in tmp.withIndex()) {
            t[i][j] = v.toInt()
        }
    }
    println(getTheLongest(t, n, 0, 0, rec))
}

fun getTheLongest(t: Array<Array<Int>>, n: Int, i: Int, j: Int, rec: Array<Array<Int>>): Int {
    if (i == n - 1) {
        return t[i][j]
    }
    if (rec[i][j] == -1) {
        rec[i][j] = max(getTheLongest(t, n, i + 1, j, rec), getTheLongest(t, n, i + 1, j + 1, rec)) + t[i][j]
    }
    return rec[i][j]
}

File: lesson-2_recursion-and-recurrence/the-longest-path/memoization/main.mjs
import { stdin as input, stdout as output } from 'node:process'
import * as readline from 'node:readline/promises'

const rl = readline.createInterface({ input, output })

const n = await rl.question('')
const t = Array.from({ length: n }, () => []),
    rec = Array.from({ length: n }, () => [])
for (let i = 0; i < n; i++) {
    const tmp = (await rl.question('')).split(' ')
    rec[i] = Array.from({ length: n }, () => -1)
    for (const [j, v] of tmp.entries()) {
        t[i][j] = Number(v)
    }
}
console.log(getTheLongest(0, 0))
rl.close()

function getTheLongest(i, j) {
    if (i === n - 1) {
        return t[i][j]
    }

    if (rec[i][j] === -1) {
        rec[i][j] = Math.max(getTheLongest(i + 1, j, rec), getTheLongest(i + 1, j + 1, rec)) + t[i][j]
    }
    return rec[i][j]
}
